testing/time.ts

```diff
@@ -0,0 +1,436 @@
+  /**
+   * The frequency in milliseconds at which fake time is updated.
+   * If advanceRate is set, we will update the time every 10 milliseconds by default.
+   */
+  advanceFrequency?: number;
+}
+
+interface DueNode {
+  due: number;
+  timers: Timer[];
+}
+
+let time: FakeTime | undefined = undefined;
+
+function fakeSetTimeout(
+  // deno-lint-ignore no-explicit-any
+  callback: (...args: any[]) => void,
+  delay = 0,
+  // deno-lint-ignore no-explicit-any
+  ...args: any[]
+): number {
+  if (!time) throw new TimeError("no fake time");
+  return setTimer(callback, delay, args, false);
+}
+
+function fakeClearTimeout(id?: number): void {
+  if (!time) throw new TimeError("no fake time");
+  if (typeof id === "number" && dueNodes.has(id)) {
+    dueNodes.delete(id);
+  }
+}
+
+/** The fake time replacement for se */
+function fakeSetInterval(
+  // deno-lint-ignore no-explicit-any
+  callback: (...args: any[]) => unknown,
+  delay = 0,
+  // deno-lint-ignore no-explicit-any
+  ...args: any[]
+): number {
+  if (!time) throw new TimeError("no fake time");
+  return setTimer(callback, delay, args, true);
+}
+
+function fakeClearInterval(id?: number): void {
+  if (!time) throw new TimeError("no fake time");
+  if (typeof id === "number" && dueNodes.has(id)) {
+    dueNodes.delete(id);
+  }
+}
+
+function setTimer(
+  // deno-lint-ignore no-explicit-any
+  callback: (...args: any[]) => void,
+  delay = 0,
+  args: unknown[],
+  repeat = false,
+): number {
+  const id: number = timerId.next().value;
+  delay = Math.max(repeat ? 1 : 0, Math.floor(delay));
+  const due: number = now + delay;
+  let dueNode: DueNode | null = dueTree.find({ due } as DueNode);
+  if (dueNode === null) {
+    dueNode = { due, timers: [] };
+    dueTree.insert(dueNode);
+  }
+  dueNode.timers.push({
+    id,
+    callback,
+    args,
+    delay,
+    due,
+    repeat,
+  });
+  dueNodes.set(id, dueNode);
+  return id;
+}
+
+function overrideGlobals(): void {
+  globalThis.Date = FakeDate as DateConstructor;
+  globalThis.setTimeout = fakeSetTimeout;
+  globalThis.clearTimeout = fakeClearTimeout;
+  globalThis.setInterval = fakeSetInterval;
+  globalThis.clearInterval = fakeClearInterval;
+}
+
+function restoreGlobals(): void {
+  globalThis.Date = _internals.Date;
+  globalThis.setTimeout = _internals.setTimeout;
+  globalThis.clearTimeout = _internals.clearTimeout;
+  globalThis.setInterval = _internals.setInterval;
+  globalThis.clearInterval = _internals.clearInterval;
+}
+
+function* timerIdGen() {
+  let i = 1;
+  while (true) yield i++;
+}
+
+let startedAt: number;
+let now: number;
+let initializedAt: number;
+let advanceRate: number;
+let advanceFrequency: number;
+let advanceIntervalId: number | undefined;
+let timerId: Generator<number>;
+let dueNodes: Map<number, DueNode>;
+let dueTree: RBTree<DueNode>;
+
+/**
+ * Overrides the real Date object and timer functions with fake ones that can be
+ * controlled through the fake time instance.
```

Comment:
Could you add an example in a codeblock?

Reply:
Sure

Reply:
Would the example from the README.md file be fine?

Reply:
Yeah, definitely!

Reply:
I've added the example from the README.md file.